pkg/controller/job/job_controller.go

```diff
@@ -1361,18 +1361,33 @@ func jobSuspended(job *batch.Job) bool {
 	return job.Spec.Suspend != nil && *job.Spec.Suspend
 }
 
+// Wait for Pods to terminate
+func waitForPodsToTerminate(job *batch.Job) bool {
+	if job.Spec.PodFailurePolicy == nil {
+		return false
+	}
+	return job.Spec.PodFailurePolicy.DelayPodRecreationUntilTerminal != nil && *job.Spec.PodFailurePolicy.DelayPodRecreationUntilTerminal
+}
+
 // manageJob is the core method responsible for managing the number of running
 // pods according to what is specified in the job.Spec.
 // Respects back-off; does not create new pods if the back-off time has not passed
 // Does NOT modify <activePods>.
-func (jm *Controller) manageJob(ctx context.Context, job *batch.Job, activePods []*v1.Pod, succeeded int32, succeededIndexes []interval, backoff backoffRecord) (int32, string, error) {
+func (jm *Controller) manageJob(ctx context.Context, job *batch.Job, allPods []*v1.Pod, activePods []*v1.Pod, succeeded int32, succeededIndexes []interval, backoff backoffRecord) (int32, string, error) {
```

Comment:
what about an alternative implementation when we include in `activePods` these pods which are still terminating, when the flag is enabled: https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/pkg/controller/job/job_controller.go#L749? 

I think it would simplify the solution conceptually, but might be missing something. Could you please try and see if anything breaks?

Reply:
I pushed up this change and it is much cleaner.  Does this resolve your concerns about IndexedJobs that you raised above?

Reply:
This was a good change! I saw that replica sets use the same code to filter active pods.

did a quick test and if we add terminating pods to be active we can also apply a similar fix to replicaset.

Obviously, we need a KEP for deployment api that is different then jobs. But at least implementation is similar